### PR TITLE
update for iOS15 signature

### PR DESCRIPTION
**Pull request by Xkinda**

### CODE REVIEW

**Comment by gb160** [comment]
Not working for me buddy, on iOS 15 I get :
`The app cannot be installed because its integrity could not be verified`

**Comment by xxqy123** [comment]
Not working for me too!  On iOS15 i got:`Application is missing the application-identifier entitlement.` while all going well on iOS14

**Comment by Xkinda** [comment]
Try signing without cache using "-f"

**Comment by gb160** [comment]
OK, I can confirm this is now working for me using iOS 15 beta, the issue I had was caused by something else.
I appreciate your work on fixing this, thanks.

**Comment by G5t4r** [comment]
iOS 15 beta not work still~:
Install: CreatingStagingDirectory (5%)
Install: ExtractingPackage (15%)
Install: InspectingPackage (20%)
Install: TakingInstallLock (20%)
Install: PreflightingApplication (30%)
Install: InstallingEmbeddedProfile (30%)
Install: VerifyingApplication (40%)ERROR: Install failed. Got error "ApplicationVerificationFailed" with code 0x00000000: Application is missing the application-identifier entitlement

**Comment by vkedwardli** [comment]
Thanks a lot, how did you find out the latest requirement? There is so little information about the EMBEDDED_ENTITLEMENTS_DER blob `0xfade7172` on the internet

**Comment by xxqy123** [comment]
> iOS 15 beta not work still~:
> Install: CreatingStagingDirectory (5%)
> Install: ExtractingPackage (15%)
> Install: InspectingPackage (20%)
> Install: TakingInstallLock (20%)
> Install: PreflightingApplication (30%)
> Install: InstallingEmbeddedProfile (30%)
> Install: VerifyingApplication (40%)ERROR: Install failed. Got error "ApplicationVerificationFailed" with code 0x00000000: Application is missing the application-identifier entitlement

Same with you. Have you found a solution?

**Comment by davidmdt21** [comment]
> iOS 15 beta not work still~:
> Install: CreatingStagingDirectory (5%)
> Install: ExtractingPackage (15%)
> Install: InspectingPackage (20%)
> Install: TakingInstallLock (20%)
> Install: PreflightingApplication (30%)
> Install: InstallingEmbeddedProfile (30%)
> Install: VerifyingApplication (40%)ERROR: Install failed. Got error "ApplicationVerificationFailed" with code 0x00000000: Application is missing the application-identifier entitlement

also i get this error when installing on iOS 15 Beta :) 
anybody can fix or help ? :-s @zhlynn

**Comment by yonglam** [comment]
@Xkinda had done a great job, but there was something wrong when deal with the blob length. 
`if (this->blobsLength > UCHAR_MAX)` will cause the `Application is missing the application-identifier entitlement` problem when the length of your `blob 0xfade7172` is larger than 127.  This is the main cause. You should change `UCHAR_MAX` to 127.

There are other similar problems when deal with `length`，for example, `put(data, (char*)& length, 1);`, because length maybe larger than 127.  you should fix it the way just like `blobsLength`.

Why is 127 and not 255? You can find the answer in [the ldid related fix](https://git.saurik.com/ldid.git/blobdiff/2edb2a9307f1bd3909dadc20e80857c6e40c00c5..98bc517f582c91e6c0577f4dfdff33753f2876e5:/ldid.cpp).

After fixed the `length` problem, I made my app running on iOS 15 perfectly. @davidmdt21 @G5t4r FYI.

**Comment by vkedwardli** [comment]
@yonglam please mark the corresponding lines with the review feature if possible 🙇‍♂️

File: archo.cpp [hunk 1]
```diff
@@ -2,9 +2,239 @@
...
+    {
+        return this->length + 2;
+    }
+
+    std::string data()
+    {
+        std::stringbuf data;
+
+        put(data, (char*)& this->type, 1);
+        put(data, (char*)& this->length, 1);
```
Comment (yonglam) [inline archo.cpp:39]: if this->length is larger than 127, we should put the value `x80 | bytesof(this->length)` in one byte which is followed by `bytesof(length)` bytes of `length`

**Comment by yonglam** [comment]
> @yonglam please mark the corresponding lines with the review feature if possible 🙇‍♂️
I will try to make a pull request when I have time.

**Comment by zhlynn** [comment]
> 
> 
> > @yonglam please mark the corresponding lines with the review feature if possible 🙇‍♂️
> > I will try to make a pull request when I have time.

expecting.  i'm very busy now.

**Comment by guthubah** [comment]
> Not working for me buddy, on iOS 15 I get :
> 
> `The app cannot be installed because its integrity could not be verified`

@gb160 i'm facing the same error, how did you fix it? I have no idea what's causing this.

**Comment by davidmdt21** [comment]
really ? no body can update Zsign for iOS 15 Beta ?:| 
we waiting from 1 month ago !   dear @zhlynn  please get some time for finish this 👍

**Comment by vkedwardli** [comment]
@davidmdt21 really? did you pay for this tool?
just use this pull request

**Comment by gb160** [comment]
> @davidmdt21 really? did you pay for this tool?
> just use this pull request

Exactly, the entitlement shown from some people these days is laughable.

**Comment by yonglam** [comment]
Have made a new PR #154 , :) @vkedwardli @zhlynn

**Comment by vkedwardli** [comment]
#172 
@Xkinda @yonglam seems it is broken on 15.1 beta again

**Pull request merged**